browser/search_engines/search_engine_tracker.cc

```diff
@@ -0,0 +1,103 @@
+// Not all engines added by Brave are present in |SearchEngineType| enumeration.
+void RecordSearchEngineP3A(const GURL& search_engine_url,
+                           SearchEngineType type) {
+  SearchEngineP3A answer = SearchEngineP3A::kOther;
+
+  if (type == SEARCH_ENGINE_GOOGLE) {
+    answer = SearchEngineP3A::kGoogle;
+  } else if (type == SEARCH_ENGINE_DUCKDUCKGO) {
+    answer = SearchEngineP3A::kDuckDuckGo;
+  } else if (type == SEARCH_ENGINE_BING) {
+    answer = SearchEngineP3A::kBing;
+  } else if (type == SEARCH_ENGINE_QWANT) {
+    answer = SearchEngineP3A::kQwant;
+  } else if (type == SEARCH_ENGINE_OTHER) {
+    if (base::EndsWith(search_engine_url.host(), "startpage.com",
+                       base::CompareCase::INSENSITIVE_ASCII)) {
+      answer = SearchEngineP3A::kStartpage;
+    }
+  }
+
+  UMA_HISTOGRAM_ENUMERATION("Brave.Search.DefaultEngine", answer);
+}
+
+}  // namespace
+
+// static
+SearchEngineTrackerFactory* SearchEngineTrackerFactory::GetInstance() {
+  return base::Singleton<SearchEngineTrackerFactory>::get();
+}
+
+SearchEngineTrackerFactory::SearchEngineTrackerFactory()
+    : BrowserContextKeyedServiceFactory(
+          "SearchEngineTracker",
+          BrowserContextDependencyManager::GetInstance()) {
+  DependsOn(TemplateURLServiceFactory::GetInstance());
+}
+
+SearchEngineTrackerFactory::~SearchEngineTrackerFactory() {}
+
+KeyedService* SearchEngineTrackerFactory::BuildServiceInstanceFor(
+    content::BrowserContext* context) const {
+  auto* template_url_service = TemplateURLServiceFactory::GetForProfile(
+      Profile::FromBrowserContext(context));
+  if (template_url_service) {
+    return new SearchEngineTracker(template_url_service);
+  }
+  return nullptr;
+}
+
+bool SearchEngineTrackerFactory::ServiceIsCreatedWithBrowserContext() const {
+  return true;
+}
+
+SearchEngineTracker::SearchEngineTracker(
+    TemplateURLService* template_url_service)
+    : template_url_service_(template_url_service) {
+  observer_.Add(template_url_service_);
+  const TemplateURL* template_url =
+      template_url_service_->GetDefaultSearchProvider();
+
+  // Record the initial P3A.
+  if (template_url) {
+    const SearchTermsData& search_terms =
+        template_url_service_->search_terms_data();
+
+    const GURL url = template_url->GenerateSearchURL(search_terms);
+    if (!url.is_empty()) {
+      default_search_url_ = url;
+      RecordSearchEngineP3A(url, template_url->GetEngineType(search_terms));
+    }
+  }
+}
+
+SearchEngineTracker::~SearchEngineTracker() {}
+
+void SearchEngineTracker::OnTemplateURLServiceChanged() {
+  const TemplateURL* template_url =
+      template_url_service_->GetDefaultSearchProvider();
+  if (template_url) {
+    const SearchTermsData& search_terms =
+        template_url_service_->search_terms_data();
+    const GURL& url = template_url->GenerateSearchURL(search_terms);
+    if (url != default_search_url_) {
```

Comment:
Just out of curiosity, this callback is called  in `url == default_search_url_` case?

Reply:
I think yes, because other details (e.g. shortcut or anything else) could be changed, and in this case the callback fires as well